python/dgl/distributed/partition.py

```diff
@@ -213,6 +213,82 @@ def load_partition_book(part_config, part_id, graph=None):
         return BasicPartitionBook(part_id, num_parts, node_map, edge_map, graph), \
                 part_metadata['graph_name'], ntypes, etypes
 
+def _get_orig_ids(g, sim_g, reshuffle, orig_nids, orig_eids):
+    '''Convert/construct the original node IDs and edge IDs.
+
+    It handles multiple cases:
+     * If the graph has been reshuffled and it's a homogeneous graph, we just return
+       the original node IDs and edge IDs in the inputs.
+     * If the graph has been reshuffled and it's a heterogeneous graph, we need to
+       split the original node IDs and edge IDs in the inputs based on the node types
+       and edge types.
+     * If the graph is not shuffled, the original node IDs and edge IDs don't change.
+
+    Parameters
+    ----------
+    g : DGLGraph
+       The input graph for partitioning.
+    sim_g : DGLGraph
+        The homogeneous version of the input graph.
+    reshuffle : bool
+        Whether the input graph is reshuffled during partitioning.
+    orig_nids : tensor or None
+        The original node IDs after the input graph is reshuffled.
+    orig_eids : tensor or None
+        The original edge IDs after the input graph is reshuffled.
+
+    Returns
+    -------
+    tensor or dict of tensors, tensor or dict of tensors
+    '''
+    is_hetero = len(g.etypes) > 1 or len(g.ntypes) > 1
+    if reshuffle and is_hetero:
+        # Get the type IDs
+        orig_ntype = F.gather_row(sim_g.ndata[NTYPE], orig_nids)
+        orig_etype = F.gather_row(sim_g.edata[ETYPE], orig_eids)
+        # Mapping between shuffled global IDs to original per-type IDs
+        orig_nids = F.gather_row(sim_g.ndata[NID], orig_nids)
+        orig_eids = F.gather_row(sim_g.edata[EID], orig_eids)
+        orig_nids = {ntype: F.boolean_mask(orig_nids, orig_ntype == g.get_ntype_id(ntype)) \
+                for ntype in g.ntypes}
+        orig_eids = {etype: F.boolean_mask(orig_eids, orig_etype == g.get_etype_id(etype)) \
+                for etype in g.etypes}
+    elif not reshuffle and not is_hetero:
+        orig_nids = F.arange(0, sim_g.number_of_nodes())
+        orig_eids = F.arange(0, sim_g.number_of_edges())
+    elif not reshuffle:
+        orig_nids = {ntype: F.arange(0, g.number_of_nodes(ntype)) for ntype in g.ntypes}
+        orig_eids = {etype: F.arange(0, g.number_of_edges(etype)) for etype in g.etypes}
+    return orig_nids, orig_eids
+
+def _set_trainer_ids(g, sim_g, node_parts):
+    '''Set the trainer IDs for each node and edge on the input graph.
+
+    The trainer IDs will be stored as node data and edge data in the input graph.
+
+    Parameters
+    ----------
+    g : DGLGraph
+       The input graph for partitioning.
+    sim_g : DGLGraph
+        The homogeneous version of the input graph.
+    node_parts : tensor
+        The node partition ID for each node in `sim_g`.
+    '''
+    if len(g.etypes) == 1:
+        g.ndata['trainer_id'] = node_parts
+        g.edata['trainer_id'] = node_parts[g.edges()[1]]
```

Comment:
Can you add some doc about how trainer_id is decided here?
From my understanding according to the code here. the trainer_id is decided according the dest node.